nomt/src/bitbox/mod.rs

```diff
@@ -256,55 +256,205 @@ impl PageLoader {
-                    }
+        let command = IoCommand {
+            kind: IoKind::Read(ht_fd, data_page_index, Box::new(Page::zeroed())),
+            user_data,
+        };
+
+        match self.io_handle.try_send(command) {
+            Ok(()) => {
+                load.state = PageLoadState::Submitted(bucket);
+                Ok(PageLoadAdvance::Submitted)
+            }
+            Err(TrySendError::Full(_)) => {
+                load.state = PageLoadState::Blocked(bucket);
+                Ok(PageLoadAdvance::Blocked)
+            }
+            Err(TrySendError::Disconnected(_)) => anyhow::bail!("I/O pool hangup"),
+        }
+    }
+
+    /// Advance the state of the given page load, blocking the current thread.
+    /// Fails if the I/O pool is down.
+    ///
+    /// Panics if the page load needs a completion.
+    ///
+    /// This returns `Ok(true)` if the page request has been submitted and a completion will be
+    /// coming. `Ok(false)` means that the page is guaranteed to be fresh.
+    pub fn advance(
+        &self,
+        ht_fd: RawFd,
+        load: &mut PageLoad,
+        user_data: u64,
+    ) -> anyhow::Result<bool> {
+        let bucket = match load.take_blocked() {
+            Some(bucket) => bucket,
+            None => loop {
+                match load.probe_sequence.next(&self.meta_map) {
+                    ProbeResult::Tombstone(_) => continue,
+                    ProbeResult::Empty(_) => return Ok(false),
+                    ProbeResult::PossibleHit(bucket) => break BucketIndex(bucket),
+                }
+            },
+        };
+
+        let data_page_index = self.shared.store.data_page_index(bucket.0);
+
+        let command = IoCommand {
+            kind: IoKind::Read(ht_fd, data_page_index, Box::new(Page::zeroed())),
+            user_data,
+        };
+
+        match self.io_handle.send(command) {
+            Ok(()) => {
+                load.state = PageLoadState::Submitted(bucket);
+                Ok(true)
+            }
+            Err(_) => anyhow::bail!("I/O pool hangup"),
+        }
+    }
+
+    /// Try to receive the next completion, without blocking the current thread.
+    ///
+    /// Fails if the I/O pool is down or a request caused an I/O error.
+    pub fn try_complete(&self) -> anyhow::Result<Option<PageLoadCompletion>> {
+        match self.io_handle.try_recv() {
+            Ok(completion) => {
+                completion.result?;
+                match completion.command.kind {
+                    IoKind::Read(_, _, page) => Ok(Some(PageLoadCompletion {
+                        page,
+                        user_data: completion.command.user_data,
+                    })),
+                    _ => panic!(),
+                }
+            }
+            Err(TryRecvError::Empty) => Ok(None),
+            Err(TryRecvError::Disconnected) => anyhow::bail!("I/O pool hangup"),
+        }
+    }
+
+    /// Receive the next completion, blocking the current thread.
```

Comment:
Or blocking forever if in incorrect state, AFAIU.

Reply:
True